3rdparty/mbedtls/mbedtls_hardware_poll.c

```diff
@@ -3,9 +3,83 @@
 
 #include <openenclave/enclave.h>
 #include <openenclave/internal/entropy.h>
+#include <string.h>
+#include "mbedtls/include/mbedtls/platform.h"
+#include "mbedtls/include/mbedtls/sha512.h"
+
+/* Per https://software.intel.com/en-us/articles/intel-digital-random-number
+ * -generator-drng-software-implementation-guide, reading 512 x 128-bit values
+ * causes RDRAND to reseed */
+#define RDRAND_RESEED_SIZE_BYTES 8192
+
+#define SHA512_HASH_LENGTH_BYTES 64
 
 int mbedtls_hardware_poll(void*, unsigned char*, size_t, size_t*);
 
+static void _fill_buffer(
+    uint8_t* source,
+    size_t source_size,
+    uint8_t** target,
+    size_t* target_size)
+{
+    size_t copy_size =
+        (*target_size > source_size) ? source_size : *target_size;
+    memcpy(*target, source, copy_size);
+    *target += copy_size;
+    *target_size -= copy_size;
+}
+
+static int _get_seed_from_rdrand(uint8_t** seed, size_t* seed_size)
+{
+    int ret = -1;
+    uint8_t* rdrand_seed = NULL;
+    uint8_t* rdrand_bytes = NULL;
+    oe_entropy_kind_t kind = OE_ENTROPY_KIND_NONE;
+
+    /* Per Intel's DRNG software implementation guide we try to obtain an
+     * equivalent amount of entropy by condensing several reseed windows of
+     * RDRAND into a single value. Reading 3x of the RDRAND_RESEED_SIZE_BYTES
+     * window provides enough entropy to be hashed back down into a single
+     * 512-bit value (64 bytes), which is typically how much entropy is
+     * requested.
+     */
+    size_t rdrand_bytes_size = RDRAND_RESEED_SIZE_BYTES * 3;
```

Comment:
Shouldn't this be 4x to ensure future calls will also have enough entropy? For example, two consecutive calls to this function might only do 7 RDSEED intervals since 6 * 512 < 7 * 511 instead of the expected 8.

Reply:
I'm not a cryptographer, so I went with the recommendation of Niels Ferguson who owns the https://github.com/microsoft/SymCrypt library. I'm not sure if it's necessary for the RDSEED intervals to be disjoint between consecutive calls as suggested.